#!/usr/bin/env/python
"""Tests for analysis.read_editor"""

import collections
import pysam
import tempfile
import unittest

import analysis.read_editor as ed
import core_utils.file_utils as fu
from core_utils.vcf_utils import get_variant_type
from definitions import *
from analysis.seq_utils import sort_and_index, COORD_FORMAT, DEFAULT_MAPQ, Strand, ReadMate, MASKED_BQ

tempfile.tempdir = os.getenv("SCRATCH", "/tmp")

TEST_VCF = """##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##contig=<ID=ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|>
##INFO=<ID=AAM_AA_CHANGE,Number=.,Type=String,Description="Can_Cenik_lab.Ian_Hoskins.bioinfo.analysis.coordinate_mapper.AminoAcidMapper comma-delimited amino acid changes.">
##INFO=<ID=VARTYPE,Number=.,Type=String,Description="Variant type. MNPs are confined to a codon. Haplotypes span codons.">
##INFO=<ID=AF,Number=1,Type=Float,Description="Allele frequency in range (0,1).">
##INFO=<ID=IE,Number=1,Type=String,Description="Introduce sequencing error to a specific strand, either + or -.">
##INFO=<ID=IR,Number=1,Type=String,Description="Introduce error (sample strand chemical change) to a specific read mate, either R1 or R2.">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	626	.	C	T	.	.	VARTYPE=snp;AF=1.0;IR=R1
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	795	.	GAC	ATG	.	.	AAM_AA_CHANGE=p.D179M;VARTYPE=tri_nt_MNP;AF=1.0
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	804	.	C	G	.	.	AAM_AA_CHANGE=p.R182G;VARTYPE=snp;AF=0.25
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	804	.	C	T	.	.	AAM_AA_CHANGE=p.R182W;VARTYPE=snp;AF=0.25
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	813	.	GG	AT	.	.	AAM_AA_CHANGE=p.G185M;VARTYPE=di_nt_MNP;AF=0.5
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	955	.	ACT	A	.	.	VARTYPE=del;AF=1.0
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1094	.	T	TG	.	.	VARTYPE=ins;AF=1.0
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1861	.	T	G	.	.	VARTYPE=snp;AF=1.0;IE=+
"""

TEST_SAM = """@HD	VN:1.0	SO:coordinate
@SQ	SN:ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	LN:2605
0000005815	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	241	42	148M	=	241	-148	ACCATCTGTCCGGTCCCAGCATGCCTTCTGAGACCCCCCAGGCAGAAGTGGGGCCCACAGGCTGCCCCCACCGCTCAGGGCCACACTCGGCGAAGGGGAGCCTGGAGAAGGGGTCCCCAGAGGATAAGGAAGCCAAGGAGCCCCTGTG	SSSSSSSSSSLOKROLRPMKIQQPJRIQJSNLIINKMSJMQNIINKSPQNNLNQNPNPJPRSOPSKPLOPOIPMSSPLOINSJQMMQMKLSNKQLIPIQRNROKPRKMJKIKRLIMSJMRRRPRRRPQNLOIKSJMIRSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:148	YS:i:0	YT:Z:CP
0000005815	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	241	42	148M	=	241	-148	ACCATCTGTCCGGTCCCAGCATGCCTTCTGAGACCCCCCAGGCAGAAGTGGGGCCCACAGGCTGCCCCCACCGCTCAGGGCCACACTCGGCGAAGGGGAGCCTGGAGAAGGGGTCCCCAGAGGATAAGGAAGCCAAGGAGCCCCTGTG	SSSSSSSSSSINKKNIRJRIJONJRNJIRQISRQIMIJMSLQJMIRRMISPKPQJJMIOQKQOQJKPIMPKJPMSNNSILPILKNRRLIJPQKMMKQIOLPSMIMKLRSILNIMNPISOPPRLRQSNRPMQPJJMRRRSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:148	YS:i:0	YT:Z:CP
0000253933	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	566	42	150M	=	569	153	GTGTGAGCTCTTGGCCAAGTGTGAGTTCTTCAACGCGGGCGGGAGCGTGAAGGACCGCATCAGCCTGCGGATGATTGAGGATGCTGAGCGCGACGGGACGCTGAAGCCCGGGGACACGATTATCGAGCCGACATCCGGGAACACCGGGAT	SSSSSSSSSSNJMMKKPJQMSSSMINMMKOSQKPKINROPPORPQLKMMIJNSSORRIIPRKRIMIPOMOPMJSNLIQQLORJRJILRJMLPSILQSKMLKKSLJPNMMQIMLKRNQMNSPKPOPQIIINIPMOSRIKROPSLSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000253933	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	569	42	150M	=	566	-153	TGAGCTCTTGGCCAAGTGTGAGTTCTTCAACGCGGGCGGGAGCGTGAAGGACCGCATCAGCCTGCGGATGATTGAGGATGCTGAGCGCGACGGGACGCTGAAGCCCGGGGACACGATTATCGAGCCGACATCCGGGAACACCGGGATCGG	SSSSSSSQIKKSQINKPSPSLRILIONLIMNKJNQPRINPOJOKONRPSKSQININOKOPONIPINNOIRSPOPJSSQPMPPLOOSKNOSJRINIKJNOMRLSSKJLLJLONQOJSMNPMJLMIMMPLSIIMMLLLONMSSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000224875	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	680	42	150M	=	683	153	CACGATTATCGAGCCGACATCCGGGAACACCGGGATCGGGCTGGCCCTGGCTGCGGCAGTGAGGGGCTATCGCTGCATCATCGTGATGCCAGAGAAGATGAGCTCCGAGAAGGTGGACGTGCTGCGGGCACTGGGGGCTGAGATTGTGAG	SSSSSSSSSSOKROSNOMKISLLKKLNPLIJQOSJJMRLQIJJQRSRNMKLLOLJSJRNJRSPPPLMRQKQMIRQRLJOPSOJMKLRRMJIJOOJQLSPLKKMJSNNQLOMJSOJKRRMPPLOOIQPNRQJSLJSOKOKMMMMSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000224875	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	683	42	150M	=	680	-153	GATTATCGAGCCGACATCCGGGAACACCGGGATCGGGCTGGCCCTGGCTGCGGCAGTGAGGGGCTATCGCTGCATCATCGTGATGCCAGAGAAGATGAGCTCCGAGAAGGTGGACGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGAC	SSSSSSSNSRKQNMLQIKNLNJPKLIINNNQRNIROIQJMJLLKOQSQIIKSOMOOQQIRRRNNMLIOLKSJQQJKKLQRSIQPLJLRMLPKOPMNPLIRLLIJNQLKISPOSKKIMIIKSMKMJKSKKORKISNPROOKSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000001369	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	794	42	150M	=	797	153	GGACGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGACGCCCACCAATGCCAGGTTCGACTCCCCGGAGTCACACGTGGGGGTGGCCTGGCGGCTGAAGAACGAAATCCCCAATTCTCACATCCTAGACCAGTACCGCAACGCCAGCAA	SSSSSSSSSSMOINJNKNNLLRNLPRRNNKLLNMRNNLJPMLSOINMSILMLMRLSLPNKLLRNJMSKJJIPQMJNRLQOJQLOSRORNQSQRPOPOSIPMPILKNNLPMNIQJPOQPILOOMORLPILSONMKLRLPPRIQPSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000051899	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	794	42	150M	=	797	153	GGACGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGACGCCCACCAATGCCAGGTTCGACTCCCCGGAGTCACACGTGGGGGTGGCCTGGCGGCTGAAGAACGAAATCCCCAATTCTCACATCCTAGACCAGTACCGCAACGCCAGCAA	SSSSSSSSSSKPLJJQLOKSLOMIPSIPMILSORJONIQLPKJSMJRKNIIJRPSSIOJSOOMSKQKOQMNLIOQRSQIPILKSRQPMNIRMSNKQOOJQSLMQRNJJPJOJLKRSOKSQIPSKKQPSPSLJJKSPRMNJNJLSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000225689	99	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	794	42	150M	=	797	153	GGACGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGACGCCCACCAATGCCAGGTTCGACTCCCCGGAGTCACACGTGGGGGTGGCCTGGCGGCTGAAGAACGAAATCCCCAATTCTCACATCCTAGACCAGTACCGCAACGCCAGCAA	SSSSSSSSSSPNIIJKQLRPOSSJKSMKOQKPIQKONRJNQRRPJNOQLOMPSPMSSNQRJIRPOOJIPLNSKNLMKPNSRLJLRMMJKIRPJIRKSJKMLSQOQQSJLPMSIPRLLQMIIROKOPNROSOPSKJJSPPSQIQSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000001369	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	797	42	150M	=	794	-153	CGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGACGCCCACCAATGCCAGGTTCGACTCCCCGGAGTCACACGTGGGGGTGGCCTGGCGGCTGAAGAACGAAATCCCCAATTCTCACATCCTAGACCAGTACCGCAACGCCAGCAACCC	SSSSSSSIRMRPPMKORMQQKPOQRJLQPLOKSQJOIMJJILKPKJQKNRIJJLJPRMNIRKJMJKKOJQPSKSSNJIILLLSRMPPNLIJOOJNNMMLPIOKLPSMQSQOQKPOSSQKISLQSLLRRKNSMRORRNJKOSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000051899	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	797	42	150M	=	794	-153	CGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGACGCCCACCAATGCCAGGTTCGACTCCCCGGAGTCACACGTGGGGGTGGCCTGGCGGCTGAAGAACGAAATCCCCAATTCTCACATCCTAGACCAGTACCGCAACGCCAGCAACCC	SSSSSSSNMRMMNRNPRQKRNRIOMSLPIIISLOQIOKKOOOMLOLQINPNMNNQOIMNJOQRSMPQISIIQINSJORMRROLSJQMLISKMLPMJMMSQINOOQOOSQMJMIJJLRSKPNSSSOMQPNSNKQJLIJJKKSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000225689	147	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	797	42	150M	=	794	-153	CGTGCTGCGGGCACTGGGGGCTGAGATTGTGAGGACGCCCACCAATGCCAGGTTCGACTCCCCGGAGTCACACGTGGGGGTGGCCTGGCGGCTGAAGAACGAAATCCCCAATTCTCACATCCTAGACCAGTACCGCAACGCCAGCAACCC	SSSSSSSNKMQOSOKSKJRQJJPKLKPRPOISPKNLQSSKLMLIKKPOSOQONSMILKSQISIJJMQOMNKQPQMQRIQRQQQSSJISSMSSQJILPIPMLOLQNMIMIRLQJSNKLKQJOSPRNMSPJKQJIQMKSIPSSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000003129	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	908	42	147M	=	908	-147	TTCTCACATCCTAGACCAGTACCGCAACGCCAGCAACCCCCTGGCTCACTACGACACCACCGCTGATGAGATCCTGCAGCAGTGTGATGGGAAGCTGGACATGCTGGTGGCTTCAGTGGGCACGGGCGGCACCATCACGGGCATTGC	SSSSSSSSSSOLPSKLNQOJLLNMPKLJJJNKQKOMMOKQPNLNOOLMJOMSQKIIQLKNLIPRQKNQIQRQIMJQSIQONKSIMONLJQSLMQOLNJSONKMNNLNNPKLQKPPIOMPQQQKOKPMSRPRKONMNNSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:147	YS:i:0	YT:Z:CP
0000003129	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	908	42	147M	=	908	-147	TTCTCACATCCTAGACCAGTACCGCAACGCCAGCAACCCCCTGGCTCACTACGACACCACCGCTGATGAGATCCTGCAGCAGTGTGATGGGAAGCTGGACATGCTGGTGGCTTCAGTGGGCACGGGCGGCACCATCACGGGCATTGC	SSSSSSSSSSIJJMPQKJKKINNNMKNPOQPJJNMLIRNKIRQPJNRPRSOQQMIMOIKNPLILONPKONSOOQJRSIQMIPNQSQSSNSKSOONKJJSQRIPOJQONPOQJRJSKRNMRQKRLNSQILNSQKOOIRSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:147	YS:i:0	YT:Z:CP
0000323491	99	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1016	42	150M	=	1019	153	GGCTTCAGTGGGCACGGGCGGCACCATCACGGGCATTGCCAGGAAGCTGAAGGAGAAGTGTCCTGGATGCAGGATCATTGGGGTGGATCCCGAAGGGTCCATCCTCGCAGAGCCGGAGGAGCTGAACCAGACGGAGCAGACAACCTACGA	SSSSSSSSSSNIILOQONPRQOPJOIOLNJJIOPNKQNPOKQRRSRJMPKSSPQQNJOMIJMKNIQRQPRSRIMNQNPONMRPKROKNORRRQQIISIQRONILSJKKRPSSIPKILKIMKQPNNOLJNKRIORKIMPMIPIKSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000323491	147	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1019	42	150M	=	1016	-153	TTCAGTGGGCACGGGCGGCACCATCACGGGCATTGCCAGGAAGCTGAAGGAGAAGTGTCCTGGATGCAGGATCATTGGGGTGGATCCCGAAGGGTCCATCCTCGCAGAGCCGGAGGAGCTGAACCAGACGGAGCAGACAACCTACGAGGT	SSSSSSSIRJLKSRNOKILLMKJMQKINJQPNSJIQNIJMLJMRPPRQIRRIOQLRKKNQMPISMOJNJOOKKPJOSMNMSKSOONJNKQRSLQPOSMLQMRLRRSOIPLKIJMNKLMJJMKQQKPPOKJLNLSJNLKSMSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000214705	163	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1358	42	150M	=	1361	153	GGGCCAGCGCTGCGTGGTCATTCTGCCCGACTCAGTGCGGAACTACATGACCAAGTTCCTGAGCGACAGGTGGATGCTGCAGAAGGGCTTTCTGAAGGAGGAGGACCTCACGGAGAAGAAGCCCTGGTGGTGGCACCTCCGTGTTCAGGA	SSSSSSSSSSILOKJNNLJKRORJMSOLMLQMNRSOQQIPKJRKOQKMKMRNKIMNIIRIJPNQIIPNKQRNQMIMRRJLPSIMJRNNNNPKNPPIJJSKRRRKRNOSLQQOOSPNKMOSJLNOSONNJRQONORIRPMSOMLSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000214705	83	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1361	42	150M	=	1358	-153	CCAGCGCTGCGTGGTCATTCTGCCCGACTCAGTGCGGAACTACATGACCAAGTTCCTGAGCGACAGGTGGATGCTGCAGAAGGGCTTTCTGAAGGAGGAGGACCTCACGGAGAAGAAGCCCTGGTGGTGGCACCTCCGTGTTCAGGAGCT	SSSSSSSKKMRPKISSNKMISKISRQPOJKJJQLNRKRIQPMJLPMPJLRKJIJQSKJJPSIQLQLRMQONOMLKNSQLKKOOSPLNJRIJSIIIQQRQLRJJPKRNSOKPRJNMSLSKSLMRLRLPQIQIIPIPKSOIKSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:150	YS:i:0	YT:Z:CP
0000099776	99	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1805	42	132M	=	1805	-132	GATCCAGTACCACAGCACCGGGAAGTCCAGTCAGCGGCAGATGGTGTTCGGGGTGGTCACCGCCATTGACTTGCTGAACTTCGTGGCCGCCCAGGAGCGGGACCAGAAGTGAAGTCCGGAGCGCTGGGCGGT	SSSSSSSSSSONQJOQQSQKNJSKRMNQKJLLOSJQJKLMQPLKKLSLRQLPPIPSLJSSJJMMOJINMRJJLSJMQLMPNNPSJQNNILPJKJNLNKPNLRJRSROMQRKKSRPRKIPLIQSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:132	YS:i:0	YT:Z:CP
0000099776	147	ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1805	42	132M	=	1805	-132	GATCCAGTACCACAGCACCGGGAAGTCCAGTCAGCGGCAGATGGTGTTCGGGGTGGTCACCGCCATTGACTTGCTGAACTTCGTGGCCGCCCAGGAGCGGGACCAGAAGTGAAGTCCGGAGCGCTGGGCGGT	SSSSSSSSSSINNIKMSNPPOONNOMMQIQOMRRSNKPLKLILMPKPPNMQNSLPRMRILNMIMIJNSLNJMOOQOPKRJNILIQPKQRNMOSNOQNJRKNONPSJLRRQIIMKOQKPOJMKSSSSSSSSSS	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:132	YS:i:0	YT:Z:CP
"""

TEST_PRIMERS = """ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	235	250	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	378	393	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	560	575	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	674	689	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	708	723	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	788	803	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	822	837	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	902	917	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	936	951	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1010	1025	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1044	1059	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1158	1173	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1352	1367	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1500	1515	CBS_target	.	-
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1799	1814	CBS_target	.	+
ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|	1926	1941	CBS_target	.	-
"""


class TestReadEditor(unittest.TestCase):
    """Tests for ReadEditor."""

    CBS_REF = "CBS.fa"

    @classmethod
    def setUpClass(cls):
        """Set up for TestReadEditor."""

        cls.tempdir = tempfile.mkdtemp()
        cls.test_dir = os.path.dirname(__file__)
        cls.test_data_dir = os.path.abspath(os.path.join(cls.test_dir, "..", "test_data"))
        cls.cbs_ref = os.path.join(cls.test_data_dir, cls.CBS_REF)

        with tempfile.NamedTemporaryFile(mode="w", suffix=".test.sam") as test_sam, \
                tempfile.NamedTemporaryFile(mode="w", suffix=".test.vcf", delete=False) as test_vcf, \
                tempfile.NamedTemporaryFile(mode="w", suffix=".test.primers.bed", delete=False) as test_primers:

            test_sam.write(TEST_SAM)
            fu.flush_files((test_sam,))
            cls.test_bam = sort_and_index(test_sam.name)

            test_vcf.write(TEST_VCF)
            cls.test_vcf = test_vcf.name

            test_primers.write(TEST_PRIMERS)
            cls.test_primers = test_primers.name

        cls.ed = ed.ReadEditor(
            cls.test_bam, variants=cls.test_vcf, ref=cls.cbs_ref, primers=cls.test_primers,
            output_dir=cls.tempdir, output_prefix="test_editor")

        cls.qname_lookup_rev = {zip(cls.ed.qname_lookup.values(), cls.ed.qname_lookup.keys())}

        cls.test_variant_config = ed.VARIANT_CONFIG_TUPLE(
            type=get_variant_type("C", "G"),
            contig="ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|",
            pos=804, ref="C", alt="G", af=0.5, ie=None, ir=None)

        cls.contig = "ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|"

    @classmethod
    def tearDownClass(cls):
        """Tear down for TestReadEditor."""

        fu.safe_remove((cls.tempdir, cls.test_bam, cls.test_vcf, cls.test_primers), force_remove=True)

    def test_get_variant_configs(self):
        """Test that variant configurations are extracted from the input VCF."""

        expected = [
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("C", "T"), contig=self.contig, pos=626, ref="C", alt="T", af=1.0, ie=None, ir=ReadMate("R1")),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("GAC", "ATG"), contig=self.contig, pos=795, ref="GAC", alt="ATG", af=1.0, ie=None, ir=None),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("C", "G"), contig=self.contig, pos=804, ref="C", alt="G", af=0.25, ie=None, ir=None),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("C", "G"), contig=self.contig, pos=804, ref="C", alt="T", af=0.25, ie=None, ir=None),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("GG", "AT"), contig=self.contig, pos=813, ref="GG", alt="AT", af=0.5, ie=None, ir=None),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("ACT", "A"), contig=self.contig, pos=955, ref="ACT", alt="A", af=1.0, ie=None, ir=None),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("T", "TG"), contig=self.contig, pos=1094, ref="T", alt="TG", af=1.0, ie=None, ir=None),
            ed.VARIANT_CONFIG_TUPLE(
                type=get_variant_type("C", "T"), contig=self.contig, pos=1861, ref="T", alt="G", af=1.0, ie=Strand("+"), ir=None),
        ]

        observed = self.ed._get_variant_configs()

        self.assertEqual(expected, observed)

    def test_get_edit_qnames_enough_amenable(self):
        """Tests editable read names are returned for proper allele frequencies."""

        expected = 2

        amenable_qnames = {0, 1, 2}
        total_amenable_qnames = 4
        qnames_to_edit = self.ed._get_edit_qnames(amenable_qnames, self.test_variant_config, total_amenable_qnames)
        observed = len(qnames_to_edit)

        self.assertEqual(expected, observed)

    def test_get_edit_qnames_none_amenable(self):
        """Tests that no read names are returned if none are editable at the position."""

        expected = 0
        amenable_qnames = set()
        total_amenable_qnames = 4
        qnames_to_edit = self.ed._get_edit_qnames(amenable_qnames, self.test_variant_config, total_amenable_qnames)
        observed = len(qnames_to_edit)

        self.assertEqual(expected, observed)

    def test_get_edit_qnames_bump(self):
        """Tests editable read names are returned."""

        expected = 1
        amenable_qnames = {0, 1, 2}
        total_amenable_qnames = 4

        variant_config = ed.VARIANT_CONFIG_TUPLE(
            type=get_variant_type("C", "G"),
            contig="ENST00000398165.7|ENSG00000160200.17|OTTHUMG00000086834.7|OTTHUMT00000195525.1|CBS-204|CBS|2605|protein_coding|",
            pos=804, ref="C", alt="G", af=0.1, ie=None, ir=None)

        qnames_to_edit = self.ed._get_edit_qnames(amenable_qnames, variant_config, total_amenable_qnames)
        observed = len(qnames_to_edit)

        self.assertEqual(expected, observed)

    def test_iterate_over_pileup_reads(self):
        """Tests that edit configs are appended to the edit dictionary."""

        expected = collections.defaultdict(list)

        # The qnames at POS 804 are 0000001369, 000224875, 000225689, 0000051899
        qname_alias_1 = self.qname_lookup_rev[]
        qname_alias_2 = self.qname_lookup_rev[]

        edit_key_1 = ed.EDIT_KEY_TUPLE(qname=qname_alias, mate=ReadMate(pileup_read.alignment.is_read1))

        edit_config_1 = ed.EDIT_CONFIG_TUPLE(
            contig=self.contig, pos=804, ref="C", alt="G", read_pos=pileup_read.query_position, ie=None, ir=None)

        edit_configs = collections.defaultdict(list)

        with pysam.AlignmentFile(self.ed.editor_preprocessor.edit_background, "rb") as edited_background_af:

            contig = list(self.ed.variant_configs)[0].contig
            target = COORD_FORMAT.format(contig, 804, 804)

            # We must use PileupColumns in iteration only
            # See https://github.com/pysam-developers/pysam/issues/746
            for pc in edited_background_af.pileup(
                    region=target, truncate=True, max_depth=self.ed.MAX_DP, stepper="all",
                    ignore_overlaps=False, ignore_orphans=False, min_base_quality=self.ed.MIN_BQ,
                    min_mapping_quality=DEFAULT_MAPQ):

                pc_coord = COORD_FORMAT.format(pc.reference_name, pc.reference_pos + 1, pc.reference_pos + 1)
                if pc_coord not in {804}:
                    continue

                var_indices = {i for i, e in enumerate(self.ed.variant_config_ids) if e == pc_coord}
                var_configs = [e for i, e in enumerate(self.ed.variant_configs) if i in var_indices]
                pc_ref_base = var_configs[0].ref

                amenable_qnames = [
                    self.ed._get_qname_alias(pileup_read.alignment.query_name) for pileup_read in pc.pileups
                    if not pileup_read.is_del and not pileup_read.is_refskip and
                       pileup_read.alignment.query_sequence[
                       pileup_read.query_position:pileup_read.query_position + len(pc_ref_base)] == pc_ref_base and
                       MASKED_BQ not in pileup_read.alignment.query_qualities[
                                        pileup_read.query_position:pileup_read.query_position + len(pc_ref_base)]
                ]

                amenable_qname_counter = collections.Counter(amenable_qnames)
                amenable_qnames = {qname for (qname, qname_count) in amenable_qname_counter.items() if qname_count == 2}
                total_amenable_qnames = len(amenable_qnames)

                self.ed._iterate_over_pileup_reads(
                    pc, self.test_variant_config, edit_configs, amenable_qnames, total_amenable_qnames)


    def test_get_edit_configs(self):
        """Tests for update of the edit dictionary with several variant configs."""

        pass

    def test_unmask_quals(self):
        """Tests for proper unmasking of qualities prior to write of reads."""

        pass

    def test_sort_key(self):
        """Tests proper indicies are returned for SNPs, MNPs, and InDels."""

        pass

    def test_edit(self):
        """Tests editing of a read object."""

        pass

    def test_iterate_over_reads(self):
        """Tests that multiple reads are edited."""
        pass

    def test_get_edited_read_pairs(self):
        """Tests filtering of edited reads."""
        pass
